File: Data Structures & Algorithms/partition-labels/submission-0.js
class Solution {
    /**
     * @param {string} S
     * @return {number[]}
     */
    partitionLabels(S) {
        let START = 97;
        let freq = new Array(26).fill(-1);
        for (let i = 0; i < S.length; i++) {
            freq[S.charCodeAt(i) - START] = i;
        }

        let startIdx = 0, lastIdx = 0;
        let output = [];
        for (let i = 0; i < S.length; i++) {
            lastIdx = Math.max(lastIdx, freq[S.charCodeAt(i) - START]);
            if (i == lastIdx) {
                output.push(lastIdx - startIdx + 1);
                startIdx = i + 1;
                lastIdx = i + 1;
            }
        }
        return output;
    }
}

File: Data Structures & Algorithms/find-minimum-in-rotated-sorted-array/submission-0.py
class Solution:
    def findMin(self, nums: List[int]) -> int:
        start, end = 0, len(nums) - 1
        if nums[start] < nums[end]:
            return nums[start]

        while start < end:
            mid = (start + end) // 2
            if nums[mid] > nums[start] and nums[mid] > nums[end]:
                start = mid
            else:
                end = mid
            print(start, end)

        return nums[(start + 1) % len(nums)]